script/utils/update-redirects-data.ts

```diff
@@ -0,0 +1,52 @@
+import fs from 'fs';
+import { isEqual } from 'lodash/fp';
+import path from 'path';
+import { Redirects } from '../../data/redirects/types';
+import { content } from '../../src/app/content/routes';
+import { BookWithOSWebData, LinkedArchiveTreeNode } from '../../src/app/content/types';
+import { flattenArchiveTree } from '../../src/app/content/utils';
+import { CACHED_FLATTENED_TREES, findArchiveTreeNodeById } from '../../src/app/content/utils/archiveTreeUtils';
+
+const redirectsPath = path.resolve(__dirname, '../../data/redirects/');
+
+const updateRedirectsData = async(currentBook: BookWithOSWebData, newBook: BookWithOSWebData) => {
+  if (currentBook.id !== newBook.id) {
+    throw new Error(
+      `updateRedirects requires two instances of the same book, `
+      + `but you've passed ${currentBook.id} and ${newBook.id}`);
+  }
+  const redirectsBookPath = path.resolve(redirectsPath, currentBook.id + '.json');
+  const redirects: Redirects = fs.existsSync(redirectsBookPath) ? await import(redirectsBookPath) : [];
+
+  const flatCurrentTree = flattenArchiveTree(currentBook.tree);
+
+  const formatSection = (section: LinkedArchiveTreeNode) => ({
+    bookId: currentBook.id,
+    pageId: section.id,
+    pathname: content.getUrl({ book: { slug: currentBook.slug }, page: { slug: section.slug } }),
+  });
+
+  const matchRedirect = (section: LinkedArchiveTreeNode) => isEqual(formatSection(section));
+
+  let countNewRedirections = 0;
+  for (const section of flatCurrentTree) {
+    const { slug } = findArchiveTreeNodeById(newBook.tree, section.id) || {};
+    // findArchiveTreeNodeById will cache book structure and we don't want this
+    // because we are comparing the same book with different version which may have different content
+    CACHED_FLATTENED_TREES.clear();
```

Comment:
This part wasn't originally in the script. I've added it today.
I think that if we ever used `update-content-versions-and-check-for-archived-slugs.ts` before then we weren't searching for missing pages properly

Reply:
really it would be more correct to add the version to the cache key

Reply:
Function which does the caching is that one:
```ts
function flattenArchiveTree(tree: LinkedArchiveTree): Array<LinkedArchiveTree | LinkedArchiveTreeSection> 
```
There are two things that we would have to add:
1. book version - we don't have access to this in `tree: LinkedArchiveTree`
2. archive URL - we have access only to `REACT_APP_ARCHIVE_URL` which is not enough when updating the archive URL because we need both at the same time

There are multiple functions that depend on `flattenArchiveTree` so all of them would have to accept additional `bookVersion` and `archiveUrl`. I've tried that and it is a mess, even with default parameters :D

Another solution would be to wrap all archive utils in a class with a constructor that accept these parameters, but this will affect tens of files since we will have to create archive utils instance for all of them.

Do you prefer any of these solutions?

Reply:
`flattenArchiveTree` is called recursively for every chapter - `tree` is a little bit confusing name because we have `book.tree` but in this case, it corresponds to "book tree" or "chapter"
`version` is only in the `id` itself but since it is called for chapters then it is a version of a section, not book

One more thing which we could do is change how we decide to cache tree - currently we are doing:
```
 if (process.env.NODE_ENV !== 'test') {
    CACHED_FLATTENED_TREES.set(tree.id, flattened);
 }
 ```
 
 I think I'll just change it to check if `APP_ENV` from `config` is set to `development` or `production`

Reply:
so it IS caching based on the version right? because of the version in the id parameter? so the only issue with the caching is if you're running multiple archive connections right?

Reply:
If version of a section/chapter is always updated when there is a slug change then I think the version shouldn't be a problem

Initially caching was a problem for unit tests. If I remember correctly it was because we were modifying book tree in order to test something

Caching is generally a problem when we can get two sections with the same id and the same version that are expected to have different properties (like when fetching from different archive version)

Reply:
when a book version is updated all of the page versions should be updated too. i think in testing the tree is modified for different tests and that creates issues, and yes when you have an archive upgrade the version isn't changed but the tree might be.

the changes that you're making to the environmental handling are confusing to me, i think having complicated environmentally aware logic is not ideal, and introducing an `undefined` environment for scripting and relying on that for functionality seems unclear to me.

i think it might be more clear if we change the caching line to be 

```diff
-export const CACHED_FLATTENED_TREES = new Map<string, Array<LinkedArchiveTree | LinkedArchiveTreeSection>>();  
+const CACHED_FLATTENED_TREES = new Map<string, Array<LinkedArchiveTree | LinkedArchiveTreeSection>>();  
+let cacheArchiveTrees=true
+export const disableArchiveTreeCaching = () => cacheArchiveTrees = false;
```

then
```diff
-if (process.env.NODE_ENV !== 'test') {
-  CACHED_FLATTENED_TREES.set(tree.id, flattened);
-}
+if (cacheArchiveTrees) {
+  CACHED_FLATTENED_TREES.set(tree.id, flattened);
+}
```

and call that from the beginning of your script and `/test/setup.ts`

Reply:
this makes a lot of sense, thanks

Reply:
I've pushed a change. `CACHED_FLATTENED_TREES` is still exported so we can clear it after updating titles in the tree in `content/utils.ts`